app/controllers/key_pairs_controller.rb

```diff
@@ -0,0 +1,56 @@
+class KeyPairsController < ApplicationController
+  before_action :find_compute_resource, :except => [:download]
+  before_action :find_resource, :only => [:download]
+
+  def index
+    @key_pairs = @compute_resource.get_compute_key_pairs
+  end
+
+  def download
+    @key_pair.update_attributes(audit_comment: _("%{user} Downloaded %{key} as pem file") % {user: User.current.name, key: @key_pair.name})
+    send_data @key_pair.secret, :filename => "#{@key_pair.name}.pem" if @key_pair
+  end
+
+  def recreate
+    if @compute_resource.try(:recreate)
+      process_success :success_msg => _('Successfully created'),
+                      :success_redirect => compute_resource_key_pairs_path(@compute_resource)
+    else
+      process_error :error_msg => _('Could not recreate a new key pair')
+    end
+  end
+
+  def remove
```

Comment:
also visible in the UI? or rater oops, check log files if you want to know how to handle it?